### PR TITLE
[TF C++] Added WhileContext to all while loop nodes

### PR DESCRIPTION
**Pull request by samdow**
Per discussions in #24842 and [here](https://groups.google.com/a/tensorflow.org/forum/?utm_medium=email&utm_source=footer#!topic/developers/IjuGQNcD6Kw), this adds the while context to all the nodes created by the while loop using the method Skye suggested. It also updates the one test that checks the while context.

cc @skye

### CODE REVIEW

**Review by skye — CHANGES_REQUESTED** [review]
Please update the comment at http://go/gh/tensorflow/tensorflow/blob/master/tensorflow/core/graph/graph.h#L290 as well.

File: tensorflow/cc/ops/while_loop.cc [hunk 1]
```diff
@@ -237,10 +238,11 @@ Status BuildWhileLoop(const Scope& scope, const std::vector<Output>& inputs,
...
 
-    // Set while_ctx for all exit nodes. We currently don't require knowing the
-    // while_ctx for any other nodes.
-    for (int i = 0; i < num_loop_vars; ++i) {
-      (*outputs)[i].node()->set_while_ctx(while_ctx);
+    const int final_while_id = scope.graph()->num_node_ids();
+
+    // Set while_ctx for all nodes created during the while
+    for (int id = next_id; id < final_while_id; ++id) {
+        scope.graph()->FindNodeId(id)->set_while_ctx(while_ctx);
```
[the remark this answers is not in the document]
Comment (samdow) [reply tensorflow/cc/ops/while_loop.cc]: Totally right. It's my bad! I was hitting the wrong tests. But all C++ tests pass now

**Review by skye — DISMISSED** [review]
Thanks!

File: tensorflow/cc/ops/while_loop_test.cc [hunk 2]
```diff
@@ -89,18 +89,19 @@ Status AddOneBody(const Scope& s, const std::vector<Output>& inputs,
 }
 
 TEST_F(WhileLoopTest, Basic) {
-  // Create loop: while (i < 10) i += 1
   Init(1);
+  int start_idx = scope_.graph()->num_node_ids();
```
Comment (shahzadlone) [inline tensorflow/cc/ops/while_loop_test.cc]: ```start_idx``` should be ```const```, if not muted/modified anywhere?

File: tensorflow/cc/ops/while_loop_test.cc [hunk 3]
```diff
@@ -89,18 +89,19 @@ Status AddOneBody(const Scope& s, const std::vector<Output>& inputs,
...
+  // Create loop: while (i < 10) i += 1
   CreateLoop(LessThanTenCond, AddOneBody);
 
   // Verify some output invariants
   WhileContext* while_ctx;
-  for (int i = 0; i < outputs_.size(); ++i) {
-    Node* node = outputs_[i].node();
-    ASSERT_TRUE(node->IsExit()) << "Output node " << i << ":\n"
-                                << node->DebugString();
+  int final_idx = scope_.graph()->num_node_ids();
```
Comment (shahzadlone) [inline tensorflow/cc/ops/while_loop_test.cc]: ```final_idx``` should be ```const```, if not muted/modified anywhere?

**Comment by samdow** [comment]
Sorry! I think I made the const changes like minutes after it was marked ready to pull. I only added const though and the C++ tests are still passing

**Review by skye — APPROVED** [review]

**Comment by skye** [comment]
No worries! To be honest I don't really understand how our PR magic works... I just re-approved the changes, so hopefully it'll run the tests and merge now.

**Comment by skye** [comment]
Sorry for the delay. I'm seeing internally that tensorflow/core:graph_control_flow_test (not sure why this didn't show up in the PR tests here, maybe I'm missing something). @samdow can you take a look?

**Comment by samdow** [comment]
Thanks for checking! I tried to run it locally but it seemed to pass for me:
screen shot 2019-02-07 at 2 47 01 pm
Would you be able to tell me what error it was getting? I'll also try rebasing in case the test changed

**Comment by skye** [comment]
Just FYI, I'm holding off on moving this forward until we resolve the discussion about the original email.

**Comment by rthadur** [comment]
Thank you all , closing this PR because the changes are already checked in internal system.

**Comment by skye** [comment]
@samdow do we still need this change?

**Comment by samdow** [comment]
Sorry for the delay! I just wanted to check it, but we do not need this change because @melissagrueter's PR will fix the AddGradients problem we were having

**Comment by skye** [comment]
No worries, I was just checking as well. Thanks!

**Pull request closed without merging**